core/trino-main/src/main/java/io/trino/execution/AddColumnTask.java

```diff
@@ -163,47 +164,58 @@ public ListenableFuture<Void> execute(
+
                 if (currentType instanceof RowType rowType) {
-                    List<RowType.Field> candidates = rowType.getFields().stream()
-                            // case-insensitive match
-                            .filter(rowField -> rowField.getName().isPresent() && rowField.getName().get().equalsIgnoreCase(fieldName))
-                            .collect(toImmutableList());
-                    if (fieldDepth == fieldPath.size()) {
-                        if (!candidates.isEmpty()) {
-                            if (statement.isColumnNotExists()) {
-                                return immediateVoidFuture();
-                            }
-                            throw semanticException(COLUMN_ALREADY_EXISTS, statement, "Field '%s' already exists", fieldName);
-                        }
-                        break;
+                    List<RowType.Field> candidates = getCandidates(rowType, fieldName);
+
+                    if (candidates.isEmpty()) {
+                        throw semanticException(COLUMN_NOT_FOUND, statement, "Field '%s' does not exist within %s", fieldName, rowType);
                     }
                     if (candidates.size() > 1) {
                         throw semanticException(AMBIGUOUS_NAME, statement, "Field path %s within %s is ambiguous", fieldPath, columnMetadata.getType());
                     }
-                    if (candidates.size() == 1) {
-                        RowType.Field rowField = getOnlyElement(candidates);
-                        currentType = rowField.getType();
-                        continue;
-                    }
+                    currentType = getOnlyElement(candidates).getType();
+                }
+                else {
+                    // TODO: Support array and map types
+                    throw semanticException(
+                            COLUMN_NOT_FOUND,
+                            statement,
+                            "Cannot resolve field '%s' within %s type when adding %s in %s",
+                            fieldName,
+                            currentType,
+                            String.join(".", fieldPath),
+                            columnMetadata.getType());
                 }
-                // TODO: Support array and map types
-                throw semanticException(
-                        COLUMN_NOT_FOUND,
-                        statement,
-                        "Cannot resolve field '%s' within %s type when adding %s in %s",
-                        fieldName,
-                        currentType,
-                        String.join(".", fieldPath),
-                        columnMetadata.getType());
             }
 
             String fieldName = getLast(statement.getColumn().getName().getParts());
+            if (!(currentType instanceof RowType)) {
+                throw new TrinoException(NOT_SUPPORTED, "Unsupported type: " + currentType);
+            }
```

Comment:
Ah, I missed this in my previous suggestion. Given that you had to add this, it'd be better to move this check to the `getCandidates()` method, update the method to take a Type instead of RowType, remove it from here, and remove the `if (currentType instanceof RowType)` in the loop above.

Reply:
Updated.